nsqd/topic.go

```diff
@@ -118,12 +149,30 @@ func (t *Topic) GetChannel(channelName string) *Channel {
 	return channel
 }
 
-// this expects the caller to handle locking
+// getOrCreateChannel expects the caller to handle locking
 func (t *Topic) getOrCreateChannel(channelName string) (*Channel, bool) {
 	channel, ok := t.channelMap[channelName]
 	if !ok {
+		if atomic.LoadInt32(&t.isDraining) == 1 {
+			// if this topic is draining, and there are no messages on the topic don't create a new channel
```

Comment:
I think this edge case is rare so i'd be willing to add that *but* adding that constraint would mean that if you had a topic paused and then started draining, you couldn't add channels before unpausing. It don't think that would be expected.

The need for this constraint is that we can add channels as long as we have messages because we know that the channels aren't draining yet, and propagation of draining to the channels is still yet to happen. After there are no messages on the topic a new channel would be a noop (it wouldn't get so messages) so it doesn't serve any point.